crates/holochain/src/conductor/conductor/app_connection_auth.rs

```diff
@@ -0,0 +1,243 @@
+            token.clone(),
+            TokenMeta {
+                installed_app_id,
+                expires_at,
+                single_use,
+            },
+        );
+        self.remove_expired_tokens();
+
+        (token, expires_at)
+    }
+
+    /// Authenticate a token and return the `InstalledAppId` that the token was issued for.
+    ///
+    /// If `app_id_restriction` is provided, the token will only be valid for the specified `InstalledAppId`.
+    /// This is useful when an app interface is restricted to a single app and tokens that would
+    /// otherwise be valid, are not valid for connecting to this app interface.
+    pub fn authenticate_token(
+        &mut self,
+        token: AppAuthenticationToken,
+        app_id_restriction: Option<InstalledAppId>,
+    ) -> ConductorResult<InstalledAppId> {
+        self.remove_expired_tokens();
+
+        match self.issued_tokens.entry(token) {
+            Entry::Occupied(entry) => {
+                let meta = { entry.get().clone() };
+
+                if meta.single_use {
+                    entry.remove();
+                }
+
+                if let Some(app_id_restriction) = app_id_restriction {
+                    if app_id_restriction != meta.installed_app_id {
+                        return Err(ConductorError::FailedAuthenticationError(
+                            "Attempt to use token in the context of another application"
+                                .to_string(),
+                        ));
+                    }
+                }
+
+                let app_id = meta.installed_app_id.clone();
+
+                Ok(app_id)
+            }
+            Entry::Vacant(_) => Err(ConductorError::FailedAuthenticationError(
+                "Invalid token".to_string(),
+            )),
+        }
+    }
+
+    fn remove_expired_tokens(&mut self) {
+        let current_time = SystemTime::now();
+
+        self.issued_tokens.retain(|_, meta| {
+            if let Some(expires_at) = meta.expires_at {
+                expires_at > current_time
+            } else {
+                // Always keep tokens that are set to not expire
+                true
+            }
+        });
+    }
+
+    #[cfg(test)]
+    fn age_tokens(&mut self) {
+        self.issued_tokens.iter_mut().for_each(|(_, meta)| {
+            meta.expires_at = Some(
+                SystemTime::now()
+                    .checked_sub(std::time::Duration::from_secs(10))
+                    .unwrap(),
+            );
+        });
+    }
+
+    #[cfg(test)]
+    fn get_tokens(&self) -> &HashMap<AppAuthenticationToken, TokenMeta> {
+        &self.issued_tokens
+    }
+}
+
+impl Default for AppAuthTokenStore {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug, Clone)]
+struct TokenMeta {
+    installed_app_id: InstalledAppId,
+    expires_at: Option<SystemTime>,
+    single_use: bool,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
```

Comment:
nothing async here, these can all be std tests and drop the `async`

Reply:
:S You saw NOTHING!

I'm so used to having to write functional tests that I've clearly forgotten how to write unit tests :)